FEAT #58193: l10n_ve_iot_mf

AlbertHamm-binaural opened the pull request:
Remplazando los SDK por los DLL de The Factory
1. Se Agrego una funcion para carga la libreria de TfhkaNet con el DLL (_load_dll).
2. Se cambio la forma en la que se imprime, antes se mantenia siempre la coneccion con la maquina fiscal, ahora Conecta → ejecuta → desconecta.
   * Se crearon las funciones de connect y disconnect las cuales ejecutan los metodos de las dll ("OpenFpCtrl" y "CloseFpCtrl").
3. Se creo una funcion para mandar los comandos a la maquina fiscal y validar por los log si se ejecutaron con exito (send_command).
4. Modificando la forma de obtener los datos de la maquina fiscal(numero z, numero de la maquina, numero de secuencia).
5. Se eliminan los archivos SDK
6. Acomodando mensajes  de error al intentar imprimir.

Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []

AlbertHamm-binaural committed:
FEAT #58193: l10n_ve_iot_mf

Remplazando los SDK por los DLL de The Factory
1. Se Agrego una funcion para carga la libreria de TfhkaNet con el DLL (_load_dll).
2. Se cambio la forma en la que se imprime, antes se mantenia siempre la coneccion con la maquina fiscal, ahora Conecta → ejecuta → desconecta.
   * Se crearon las funciones de connect y disconnect las cuales ejecutan los metodos de las dll ("OpenFpCtrl" y "CloseFpCtrl").
3. Se creo una funcion para mandar los comandos a la maquina fiscal y validar por los log si se ejecutaron con exito (send_command).
4. Modificando la forma de obtener los datos de la maquina fiscal(numero z, numero de la maquina, numero de secuencia).
5. Se eliminan los archivos SDK
6. Acomodando mensajes  de error al intentar imprimir.

Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []

AlbertHamm-binaural committed:
FEAT #58193: l10n_ve_iot_mf

Remplazando los SDK por los DLL de The Factory
1. Se Agrego una funcion para carga la libreria de TfhkaNet con el DLL (_load_dll).
2. Se cambio la forma en la que se imprime, antes se mantenia siempre la coneccion con la maquina fiscal, ahora Conecta → ejecuta → desconecta.
   * Se crearon las funciones de connect y disconnect las cuales ejecutan los metodos de las dll ("OpenFpCtrl" y "CloseFpCtrl").
3. Se creo una funcion para mandar los comandos a la maquina fiscal y validar por los log si se ejecutaron con exito (send_command).
4. Modificando la forma de obtener los datos de la maquina fiscal(numero z, numero de la maquina, numero de secuencia).
5. Se eliminan los archivos SDK
6. Acomodando mensajes  de error al intentar imprimir.

Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []

AlbertHamm-binaural committed:
FEAT #58193: l10n_ve_pos_mf

1. Modificando la forma de imprimir desde el pos.
2. Acomodando mensajes  de error al intentar imprimir.

Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []

AlbertHamm-binaural committed:
FEAT #58193: l10n_ve_pos_mf

1. Modificando la forma de imprimir desde el pos.
2. Acomodando mensajes  de error al intentar imprimir.

Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []

AlbertHamm-binaural committed:
manifest

AlbertHamm-binaural committed:
manifest

AlbertHamm-binaural committed:
Mejoras en validaciones y manejo de impresión fiscal

- Se agregó validación para confirmar el estado de la máquina fiscal antes de imprimir o consultar información.
- Se eliminaron loggers innecesarios.
- Se corrigió la lógica de impresión en funciones relacionadas.
- Se actualizaron funciones de consulta a la máquina fiscal (migración de SDK a DLL).
- Se ajustaron validaciones de estado y manejo de errores.
- Se modificó la forma de obtener los datos de la máquina para guardarlos en la factura.
- Se corrigió la impresión de notas de crédito desde el POS:
    - Se eliminó la función `print_out_invoice`.
    - Se adaptó `print_document` para permitir la impresión de notas de crédito.

AlbertHamm-binaural committed:
Mejoras en validaciones y manejo de impresión fiscal

- Se agregó validación para confirmar el estado de la máquina fiscal antes de imprimir o consultar información.
- Se eliminaron loggers innecesarios.
- Se corrigió la lógica de impresión en funciones relacionadas.
- Se actualizaron funciones de consulta a la máquina fiscal (migración de SDK a DLL).
- Se ajustaron validaciones de estado y manejo de errores.
- Se modificó la forma de obtener los datos de la máquina para guardarlos en la factura.
- Se corrigió la impresión de notas de crédito desde el POS:
    - Se eliminó la función `print_out_invoice`.
    - Se adaptó `print_document` para permitir la impresión de notas de crédito.

AlbertHamm-binaural committed:
Quitando logger innecesarios

AlbertHamm-binaural committed:
Quitando logger innecesarios

AlbertHamm-binaural committed:
FEAT #9703: l10n_ve_invoice

Problema: El boton cancelar debe aparecer disponible en las notas de debito y credito si forman parte del mismo período fiscal.

Solución: Cálculo de entry_in_period y control de visibilidad del botón de cancelar

- Se agregó un campo booleano computado `entry_in_period` en `account.move` para identificar si una nota de débito o crédito está dentro del período fiscal vigente.
- La lógica considera el tipo de contribuyente (especial, ordinario, formal) y la fecha de la factura en relación con el límite del período.
- Se actualizó la vista XML para incluir el campo y condicionar la visibilidad del botón de cancelar: solo se muestra si el documento está dentro del período fiscal y cumple con los criterios de estado y tipo.

Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=9703&cids=2&menu_id=302&action=386&model=helpdesk.ticket&view_type=form

Tarea de proyecto []
Ticket de soporte [x]

AlbertHamm-binaural committed:
manifest

OmarYepez29 committed:
[FIX]Se descomenta linea del cierre de la factura.

Se descomenta la linea de cierre de la factura,adicionalmente se hace
que la validacion que revisa la respuesta de la maquina fiscal ignora el
resultado en caso de que sean los comandos de cierre.

OmarYepez29 committed:
[FIX]Se descomenta linea del cierre de la factura.

Se descomenta la linea de cierre de la factura,adicionalmente se hace
que la validacion que revisa la respuesta de la maquina fiscal ignora el
resultado en caso de que sean los comandos de cierre.

AlbertHamm-binaural committed:
FIX: l10n_ve_iot_mf

Problema: La direccion esta dando error al mandar a imprimir y exede el limite

Solución: Se modifico la forma en la que se envia la direccion y telefono para que cuando la direccion supere los caracteres permitidos haga un salto de linea y el telefono tambien salte la linea para no chocar

AlbertHamm-binaural committed:
FIX: l10n_ve_iot_mf

Problema: La direccion esta dando error al mandar a imprimir y exede el limite

Solución: Se modifico la forma en la que se envia la direccion y telefono para que cuando la direccion supere los caracteres permitidos haga un salto de linea y el telefono tambien salte la linea para no chocar

AlbertHamm-binaural committed:
FIX #9703: l10n_ve_invoice, l10n_ve_stock_account

Problema: La fecha de cierre fiscal se esta calculando de una forma que no se adapta a todos los meses.

Solución: Se agrego una libreria(calendar) con la cual se puede obtener el ultimo dia del mes, con esto se modifico la forma de obtener el ultimo dia del mes actual.

Tarea (Link): https://binaural.odoo.com/web#id=9703&cids=2&menu_id=302&action=386&model=helpdesk.ticket&view_type=form

Tarea de proyecto []
Ticket de soporte [x]

AlbertHamm-binaural committed:
Campo 'mf_invoice_number' invisible en la vista

AlbertHamm-binaural committed:
Campo 'mf_invoice_number' invisible en la vista

AlbertHamm-binaural committed:
[FIX] Acomodando retorno de reimpresion

AlbertHamm-binaural committed:
FIX #58193: l10n_ve_iot_mf

Problema: Al imprimir una reimpresion sale un error que dice "Error
Desconocido".

Solución: Se acomodo la forma de retornar la respuesta del metodo que
reimprime un documento, el cual se estaba haciendo mal lo que ocacionaba
que no cumpliera con una condicion.

Tarea (Link):
https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []

AlbertHamm-binaural committed:
Merge branch '17.0' into 17.0_feat-ta_58193_dll_integration_l10n_ve_iot_mf

AlbertHamm-binaural committed:
Merge branch '17.0' into 17.0_feat-ta_58193_dll_integration_l10n_ve_iot_mf

AlbertHamm-binaural committed:
Merge branch '17.0_feat-ta_58193_dll_integration_l10n_ve_iot_mf' of github.com:binaural-dev/odoo-venezuela into 17.0_feat-ta_58193_dll_integration_l10n_ve_iot_mf

AlbertHamm-binaural committed:
Merge branch '17.0' into 17.0_feat-ta_58193_dll_integration_l10n_ve_iot_mf

AlbertHamm-binaural committed:
FIX #58193: l10n_ve_iot_mf, l10n_ve_pos_mf
CORRECCIONES:
- Agregando mensajes de validacion y quitando console.log y loggers

Remplazando los SDK por los DLL de The Factory

Se Agrego una funcion para carga la libreria de TfhkaNet con el DLL (_load_dll).
Se cambio la forma en la que se imprime, antes se mantenia siempre la coneccion con la maquina fiscal, ahora Conecta → ejecuta → desconecta.
Se crearon las funciones de connect y disconnect las cuales ejecutan los metodos de las dll ("OpenFpCtrl" y "CloseFpCtrl").
Se creo una funcion para mandar los comandos a la maquina fiscal y validar por los log si se ejecutaron con exito (send_command).
Modificando la forma de obtener los datos de la maquina fiscal(numero z, numero de la maquina, numero de secuencia).
Se eliminan los archivos SDK
Acomodando mensajes de error al intentar imprimir.
Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -218,122 +298,452 @@ def _set_actions(self):
+            
+            if estado_s1:
+                machine_number = estado_s1.RegisteredMachineNumber 
+                self.device_name = f"Impresora TFHKA: {model} - {country} - {machine_number}"
```
No usar f-string

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -218,122 +298,452 @@ def _set_actions(self):
+                machine_number = estado_s1.RegisteredMachineNumber 
+                self.device_name = f"Impresora TFHKA: {model} - {country} - {machine_number}"
+            else:
+                self.device_name = f"Impresora TFHKA: {model} - {country} - Registro desconocido"
```
No usar f-string

- [ ]

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -218,122 +298,452 @@ def _set_actions(self):
+
+            self.device_manufacturer = "The Factory HKA"
+
+            _logger.info(f" Nombre del dispositivo establecido: {self.device_name}")
```
rt

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -218,122 +298,452 @@ def _set_actions(self):
+            _logger.info(f" Nombre del dispositivo establecido: {self.device_name}")
+        
+        except Exception as e:
+            _logger.error(f"Error al establecer el nombre del dispositivo: {e}")
```
rt

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -218,122 +298,452 @@ def _set_actions(self):
+        """
+        try:
+            result = self.tfhka.SendCmd(command)
+            _logger.info(f"Comando '{command}' enviado con éxito: {result}")
```
rt

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -567,442 +987,259 @@ def filter_unique_type_method(payment):
+    
+    
+    def _print_out_refund(self, invoice):
+        """
```
docstring en ingles

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -1319,46 +1613,68 @@ def _FetchRow_Report(self, r):
-            status = self._GetStatusError(0, 144)
+
+        except Exception as e:
+            _logger.error(f"Error al leer estado fiscal: {e}")
```
f-string

OmarYepez29 commented on l10n_ve_iot_mf/iot_handlers/drivers/SerialFiscalDriver.py:
```diff
@@ -1524,97 +1840,128 @@ def _ReadFiscalMemoryByDate(self, cmd):
+            _logger.info("Datos del estado S1 obtenidos correctamente. : %s", data)
+            return data
+        except Exception as e:
+            _logger.error(f"Error al obtener los datos S1: {e}")
```
rt

OmarYepez29 commented on l10n_ve_pos_mf/static/src/js/PosState.js:
```diff
@@ -97,13 +97,25 @@ patch(PosStore.prototype, {
         if (response.length > 0) {
+          const date = new Date(response[0].date_order);
+          const format_date = date.toLocaleDateString('es-ES');
+          console.log(format_date);
```
eliminar console.log

OmarYepez29 commented on l10n_ve_pos_mf/static/src/js/PosState.js:
```diff
@@ -157,68 +169,57 @@ patch(PosStore.prototype, {
-            message: _t("Error connecting to the fiscal machine, check if it is turned on or connected to the IoT"),
-            printer_connection: false
+    } catch (err) {
+        console.log("MF error: ", err)
```
Eliminar console.log en su defecto cambiar por console.error

OmarYepez29 commented on l10n_ve_pos_mf/static/src/js/PosState.js:
```diff
@@ -244,21 +244,13 @@ patch(PosStore.prototype, {
     
     } catch (err) {
-
+      console.log("MF error: ", err)
```
rt

AlbertHamm-binaural committed:
FIX #58193: l10n_ve_iot_mf, l10n_ve_pos_mf

CORRECCIONES:
- Quitando f-string, modificando console.log, pasando a ingles docstring

Remplazando los SDK por los DLL de The Factory

Se Agrego una funcion para carga la libreria de TfhkaNet con el DLL (_load_dll).
Se cambio la forma en la que se imprime, antes se mantenia siempre la coneccion con la maquina fiscal, ahora Conecta → ejecuta → desconecta.
Se crearon las funciones de connect y disconnect las cuales ejecutan los metodos de las dll ("OpenFpCtrl" y "CloseFpCtrl").
Se creo una funcion para mandar los comandos a la maquina fiscal y validar por los log si se ejecutaron con exito (send_command).
Modificando la forma de obtener los datos de la maquina fiscal(numero z, numero de la maquina, numero de secuencia).
Se eliminan los archivos SDK
Acomodando mensajes de error al intentar imprimir.
Tarea (Link): https://binaural.odoo.com/web?db=binaural-dev-binaural-release-10413381&token=4P6sT2NxLGgos2OwxwuG#id=58193&cids=2&menu_id=975&action=341&model=project.task&view_type=form

Tarea de proyecto [x]
Ticket de soporte []
